File: scripts/garmin_sync.py
#!/usr/bin/env python3
"""
garmin_sync.py — pull last night's sleep from Garmin Connect and push it into
the somn Google Sheet (the same sheet /api/sheets reads from).

Env vars:
  GARMIN_USERNAME   (required)
  GARMIN_PASSWORD   (required)
  SHEETS_API_URL    (optional — defaults to the prod Apps Script URL)
  SOMN_NAME         (optional — defaults to "Petrica Cosmin Moga")
  SYNC_DATE         (optional — YYYY-MM-DD, defaults to yesterday in Bucharest)
  DRY_RUN           (optional — "1" to skip the POST and just print)
"""

from __future__ import annotations

import os
import sys
from datetime import date, datetime, timedelta
from urllib.parse import urlencode
from zoneinfo import ZoneInfo

import requests
from garminconnect import (
    Garmin,
    GarminConnectAuthenticationError,
    GarminConnectConnectionError,
    GarminConnectTooManyRequestsError,
)

DEFAULT_SHEETS_API = (
    "https://script.google.com/macros/s/"
    "AKfycbwNbyFuoNJV6XPAWbwANg1DOuW9rBshHlBrm3cLPcDeZORwu2L2k8N6VoHYNKdoyKhYtg/exec"
)
DEFAULT_NAME = "Petrica Cosmin Moga"
TZ = ZoneInfo("Europe/Bucharest")


def log(msg: str) -> None:
    print(f"[garmin-sync] {msg}", flush=True)


def resolve_date() -> str:
    override = os.environ.get("SYNC_DATE", "").strip()
    if override:
        # Validate format
        datetime.strptime(override, "%Y-%m-%d")
        return override
    yesterday = datetime.now(TZ).date() - timedelta(days=1)
    return yesterday.isoformat()


def login() -> Garmin:
    user = os.environ.get("GARMIN_USERNAME", "").strip()
    pwd = os.environ.get("GARMIN_PASSWORD", "")
    if not user or not pwd:
        log("ERROR: GARMIN_USERNAME / GARMIN_PASSWORD not set")
        sys.exit(2)
    client = Garmin(user, pwd)
    client.login()
    return client


def extract_sleep(raw: dict, target_date: str) -> dict | None:
    """
    Map the raw Garmin sleep payload → the fields somn cares about.
    Returns None if the night isn't logged yet.
    """
    daily = raw.get("dailySleepDTO") or {}
    if not daily:
        return None

    # Sleep score — under sleepScores.overall.value (0-100)
    scores = daily.get("sleepScores") or {}
    overall = (scores.get("overall") or {}).get("value")

    # REM minutes — remSleepSeconds → minutes (rounded)
    rem_sec = daily.get("remSleepSeconds")
    rem = round(rem_sec / 60) if isinstance(rem_sec, (int, float)) else None

    # HRV avg last night — under "hrvData" / "avgOvernightHrv" / "averageHRV"
    hrv = (
        raw.get("avgOvernightHrv")
        or (raw.get("hrvSummary") or {}).get("lastNightAvg")
        or daily.get("avgOvernightHrv")
    )

    return {
        "date": target_date,
        "ss": overall,
        "rem": rem,
        "hrv": int(hrv) if isinstance(hrv, (int, float)) else None,
    }


def fetch_rhr(client: Garmin, target_date: str) -> int | None:
    """Resting HR for the target date — try a few endpoints since the lib's API has shifted."""
    # Newer lib: get_rhr_day(date) → {"allMetrics": {"metricsMap": {"WELLNESS_RESTING_HEART_RATE": [...]}}}
    try:
        data = client.get_rhr_day(target_date)
    except Exception as e:
        log(f"get_rhr_day failed: {e}")
        data = None

    if isinstance(data, dict):
        metrics = (
            (data.get("allMetrics") or {})
            .get("metricsMap", {})
            .get("WELLNESS_RESTING_HEART_RATE", [])
        )
        for m in metrics:
            v = m.get("value")
            if isinstance(v, (int, float)) and v > 0:
                return int(v)

    # Fallback: get_stats(date)["restingHeartRate"]
    try:
        stats = client.get_stats(target_date)
        v = stats.get("restingHeartRate") if isinstance(stats, dict) else None
        if isinstance(v, (int, float)) and v > 0:
            return int(v)
    except Exception as e:
        log(f"get_stats fallback failed: {e}")

    return None


def fetch_hrv(client: Garmin, target_date: str) -> int | None:
    """Average overnight HRV in ms for the target date, if Garmin recorded it."""
    try:
        data = client.get_hrv_data(target_date)
    except Exception as e:
        log(f"get_hrv_data failed: {e}")
        return None
    if not isinstance(data, dict):
        return None
    summary = data.get("hrvSummary") or {}
    v = summary.get("lastNightAvg") or summary.get("weeklyAvg")
    if isinstance(v, (int, float)) and v > 0:
        return int(v)
    return None


def post_to_sheet(entry: dict, name: str) -> None:
    if entry.get("ss") is None or entry.get("rhr") is None:
        log(f"SKIP write — incomplete entry (ss={entry.get('ss')}, rhr={entry.get('rhr')})")
        return

    base = os.environ.get("SHEETS_API_URL", "").strip() or DEFAULT_SHEETS_API
    params = {
        "action": "write",
        "date": entry["date"],
        "name": name,
        "sleep_score": str(entry["ss"]),
        "rhr": str(entry["rhr"]),
        "hrv": "" if entry.get("hrv") is None else str(entry["hrv"]),
        "rem": "" if entry.get("rem") is None else str(entry["rem"]),
        "journal": "",
    }
    url = f"{base}?{urlencode(params)}"

    if os.environ.get("DRY_RUN") == "1":
        log(f"DRY_RUN — would GET {url}")
        return

    log(f"POST entry → {entry}")
    res = requests.get(url, timeout=30, allow_redirects=True)
    if not res.ok:
        log(f"ERROR Sheets API {res.status_code}: {res.text[:200]}")
        sys.exit(3)
    log(f"OK — Sheets API {res.status_code}")


def main() -> None:
    target_date = resolve_date()
    name = os.environ.get("SOMN_NAME", "").strip() or DEFAULT_NAME
    log(f"target_date={target_date} name={name!r}")

    try:
        client = login()
    except GarminConnectAuthenticationError as e:
        log(f"AUTH FAILED — check GARMIN_USERNAME / GARMIN_PASSWORD: {e}")
        sys.exit(2)
    except (GarminConnectConnectionError, GarminConnectTooManyRequestsError) as e:
        log(f"GARMIN UNREACHABLE: {e}")
        sys.exit(4)

    log("logged in")

    try:
        raw = client.get_sleep_data(target_date)
    except Exception as e:
        log(f"get_sleep_data failed: {e}")
        sys.exit(5)

    entry = extract_sleep(raw or {}, target_date)
    if entry is None:
        log(f"No sleep recorded for {target_date} yet — exiting cleanly.")
        return

    if entry.get("hrv") is None:
        entry["hrv"] = fetch_hrv(client, target_date)
    entry["rhr"] = fetch_rhr(client, target_date)

    log(
        f"pulled: ss={entry.get('ss')} rhr={entry.get('rhr')} "
        f"hrv={entry.get('hrv')} rem={entry.get('rem')}"
    )

    post_to_sheet(entry, name)


if __name__ == "__main__":
    main()

File: scripts/garmin-sync.mjs
/**
 * Standalone Garmin → Sheet sync, run by GitHub Actions (see
 * .github/workflows/garmin-sync.yml) on a daily cron.
 *
 * Why not the Vercel /api/garmin/sync route: @gooin/garmin-connect drags a
 * heavy transitive tree (axios → form-data → …) that Vercel's serverless
 * tracer keeps dropping. In an Actions runner `npm ci` installs the whole tree,
 * so the library just works — the same code path that passes locally.
 *
 * Env (from GitHub repo secrets):
 *   GARMIN_EMAIL, GARMIN_PASSWORD, GARMIN_PLAYER  — the account + Sheet name
 *   SHEETS_API_URL                                — Apps Script Web App URL
 *   SHEETS_TOKEN                                  — write token (gates mutations)
 *   SYNC_DAYS (optional, default 3)               — look-back window
 */

import pkg from '@gooin/garmin-connect';
const { GarminConnect } = pkg;

const {
  GARMIN_EMAIL, GARMIN_PASSWORD, GARMIN_PLAYER,
  SHEETS_API_URL, SHEETS_TOKEN, SYNC_DAYS,
} = process.env;

for (const [k, v] of Object.entries({ GARMIN_EMAIL, GARMIN_PASSWORD, GARMIN_PLAYER, SHEETS_API_URL, SHEETS_TOKEN })) {
  if (!v) { console.error(`Missing env ${k}`); process.exit(1); }
}

const DAYS = Math.min(Math.max(parseInt(SYNC_DAYS ?? '3', 10) || 3, 1), 92);

/** Normalize whatever the Sheet returns (Google coerces dates) to YYYY-MM-DD. */
function normDate(raw) {
  const s = String(raw ?? '').trim();
  if (/^\d{4}-\d{2}-\d{2}$/.test(s)) return s;
  const d = new Date(s);
  if (isNaN(d.getTime())) return s.slice(0, 10);
  return new Date(d.getTime() + 12 * 3600 * 1000).toISOString().slice(0, 10);
}

function localHHMM(tsLocal) {
  if (!tsLocal) return '';
  const d = new Date(tsLocal);
  return `${String(d.getUTCHours()).padStart(2, '0')}:${String(d.getUTCMinutes()).padStart(2, '0')}`;
}

async function existingKeys() {
  const res = await fetch(`${SHEETS_API_URL}?v=${Date.now()}`, { cache: 'no-store' });
  if (!res.ok) throw new Error(`Sheets read ${res.status}`);
  const json = await res.json();
  const set = new Set();
  for (const r of json.data ?? []) set.add(`${normDate(r.date)}::${String(r.name ?? '').trim()}`);
  return set;
}

async function writeEntry(e) {
  const p = new URLSearchParams({
    action: 'write', token: SHEETS_TOKEN,
    date: e.date, name: e.name,
    sleep_score: String(e.ss), rhr: String(e.rhr),
    hrv: e.hrv == null ? '' : String(e.hrv),
    rem: e.rem == null ? '' : String(e.rem),
    journal: '', start: e.start ?? '', end: e.end ?? '',
  });
  const res = await fetch(`${SHEETS_API_URL}?${p}`, { method: 'GET', cache: 'no-store' });
  if (!res.ok) throw new Error(`Sheets write ${res.status}`);
  const j = await res.json().catch(() => ({}));
  if (j.status === 'error') throw new Error(`Sheets write rejected: ${j.error}`);
}

async function fetchNight(client, player, date) {
  const sleep = await client.getSleepData(date);
  const dto = sleep?.dailySleepDTO;
  if (!dto?.calendarDate || !dto.sleepTimeSeconds) return null;

  const ss = dto.sleepScores?.overall?.value;
  let rhr = sleep.restingHeartRate;
  if (!rhr) {
    try { rhr = (await client.getHeartRate(date))?.restingHeartRate; } catch { rhr = 0; }
  }
  if (!ss || !rhr) return null;

  const remSec = dto.remSleepSeconds;
  const hrv = sleep.avgOvernightHrv;
  return {
    date: dto.calendarDate,
    name: player,
    ss: Math.round(ss),
    rhr: Math.round(rhr),
    hrv: hrv ? Math.round(hrv) : null,
    rem: dto.deviceRemCapable && remSec ? Math.round(remSec / 60) : null,
    start: localHHMM(dto.sleepStartTimestampLocal),
    end: localHHMM(dto.sleepEndTimestampLocal),
  };
}

async function main() {
  const existing = await existingKeys();
  const client = new GarminConnect({ username: GARMIN_EMAIL, password: GARMIN_PASSWORD });
  await client.login();

  const synced = [], skipped = [], empty = [];
  for (let i = 0; i < DAYS; i++) {
    const d = new Date(Date.now() - i * 86400000);
    const iso = d.toISOString().slice(0, 10);
    if (existing.has(`${iso}::${GARMIN_PLAYER}`)) { skipped.push(iso); continue; }
    const entry = await fetchNight(client, GARMIN_PLAYER, d);
    if (!entry) { empty.push(iso); continue; }
    if (existing.has(`${entry.date}::${entry.name}`)) { skipped.push(entry.date); continue; }
    await writeEntry(entry);
    existing.add(`${entry.date}::${entry.name}`);
    synced.push(entry.date);
  }

  console.log(JSON.stringify({ ok: true, player: GARMIN_PLAYER, synced, skipped, empty }, null, 2));
}

main().catch(err => { console.error(err); process.exit(1); });
